Make configuration work independent of whether appsettings json files are provided

ChrisDoernen opened the pull request:
Hi @Arash-Sabet,

It is me again ☺️ Thanks for merging so quick! I updated to the new version and have a follow-up suggestion.

Currently the configuration is only getting built when there are appsetting files configured (`GetTestAppSettings`). We often have the case that our configuration is only coming from environment variables without appsettings file. 

So the idea is if we can actually build the configuration regardless. This also has the advantage that the `IConfiguration` does not need to be nullable in `AddServices`. This makes it a bit nicer for the user because they don't need to check for `null`. 

I also changed `GetTestAppSettings` to be virtual with a default implementation of an empty enumerable, so users don't need to override the method when they don't want to provide appsettings. This would then work like `AddUserSecretes`.

I adopted the configuration tests (see `TestProjectFixtureWithoutAppsettings`), all test are still green.

All the best

Chris

ChrisDoernen committed:
Make configuration work independent of whether appsettings json files are provided

Arash-Sabet committed:
Merge branch 'main' into get-configuration-root-regardless

Arash-Sabet commented on examples/Xunit.Microsoft.DependencyInjection.ExampleTests/ConfigurationTests.cs:
@ChrisDoernen Can you please restore the original implementation of `ConfigurationTersts` class  and instead add a new configuration tests class to verify `TestProjectFixtureWithoutAppsettings`'s implementation?

Arash-Sabet commented on src/Abstracts/TestBedFixture.cs:
```diff
@@ -147,13 +147,10 @@ protected virtual ILoggingBuilder AddLoggingProvider(ILoggingBuilder loggingBuil
+	private IConfigurationRoot GetConfigurationRoot()
 	{
 		var testAppSettings = GetTestAppSettings();
-		return
```
@ChrisDoernen what is the reason for removing the condition?

ChrisDoernen commented:
Nice! Yes sure, take your time.

I did not yet use this feature, do you mean overriding `AddUserSecrets`? I will do research and testing on my local machine and report back.

ChrisDoernen committed:
Programatically set (random) user secrets using dotnet command. Then assert the secrets are available in the application via configuration.

ChrisDoernen committed:
Merge remote-tracking branch 'origin/get-configuration-root-regardless' into get-configuration-root-regardless

ChrisDoernen commented:
@Arash-Sabet ok I got what you mean, check out my second commit.

I am setting the (randomized) secrets in the test and assert that they are available. However, the secrets remain set after the test did run. 

For me it is not a problem and the randomization ensures no collisions. If you want me to do a clean up, I am happy to implement it.

Arash-Sabet commented on examples/Xunit.Microsoft.DependencyInjection.ExampleTests/appsettings.json:
```diff
@@ -1,9 +1,5 @@
 ﻿{
   "Options": {
     "Rate": 10
-  },
```
Hey @ChrisDoernen we did not need to remove anything or alter anything related to using secret values. Can you please undo this change? 
Let's put it in this way: I will only accept this PR if the changes are incremental. I do not want to get the already-implemented feature excluded or removed unless they are obsolete.

Arash-Sabet commented on examples/Xunit.Microsoft.DependencyInjection.ExampleTests/SecretValues.cs:
```diff
@@ -2,6 +2,6 @@
 {
-    public string? Secret1 { get; set; }
-    public string? Secret2 { get; set; }
+    public required string Secret1 { get; set; }
```
@ChrisDoernen the `required` keyword is unnecessary.

Arash-Sabet commented on examples/Xunit.Microsoft.DependencyInjection.ExampleTests/UserSecretTests.cs:
```diff
@@ -1,23 +1,45 @@
     public void TestSecretValues()
     {
-        /*
-         * TODO: Create a user secret entry like the following payload in user secrets and remove the same from appsettings.json file:
```
@ChrisDoernen Please restore this block of comment

Arash-Sabet commented on examples/Xunit.Microsoft.DependencyInjection.ExampleTests/UserSecretTests.cs:
```diff
@@ -1,23 +1,45 @@
+    {
+        SetSecret(nameof(SecretValues.Secret1), _secret1Value);
+        SetSecret(nameof(SecretValues.Secret2), _secret2Value);
+    }
```
@ChrisDoernen Please remove this block.

Arash-Sabet commented on examples/Xunit.Microsoft.DependencyInjection.ExampleTests/UserSecretTests.cs:
```diff
@@ -1,23 +1,45 @@
+        Assert.Equal(secretValues.Secret2, _secret2Value);
+    }
+
+    private void SetSecret(string secretName, string secretValue)
```
@ChrisDoernen Unnecessary.

Arash-Sabet commented:
@ChrisDoernen I left a few comments in this PR and you can review them in the "File View" section. Resolving them paves the path for my next step.

ChrisDoernen commented:
@Arash-Sabet I can not find the "File View" section and your comments. Can you post a screenshot where to click? I expected to view comments in the "Files Changed" tab:

<img width="3789" height="1590" alt="image" src="https://github.com/user-attachments/assets/23ecb6f6-8dc1-4731-b75f-89ba181d8a66" />

Arash-Sabet commented:
@ChrisDoernen Does not vs code display my comments if you open the changed files in it? It should I guess.

ChrisDoernen commented:
I am working in JetBrains Rider... 
But surely I should see them somewhere here in GitHub in the PR!?

ChrisDoernen commented:
Opened it in VSCode but nothing. Can you try commenting here on Github like so? Maybe then I see them...

<img width="1827" height="1053" alt="image" src="https://github.com/user-attachments/assets/7c3e9280-3884-4d95-9deb-63c1d641cfa5" />

Arash-Sabet commented:
@ChrisDoernen Please see the screenshots below:

<img width="3800" height="1504" alt="image" src="https://github.com/user-attachments/assets/8bdcfb97-873b-4e34-b556-31a13d5de7a8" />

<img width="3800" height="1771" alt="image" src="https://github.com/user-attachments/assets/dbc869d8-119d-49f7-b725-3814ed1dd64d" />
<img width="3800" height="793" alt="image" src="https://github.com/user-attachments/assets/d41dbfa9-6f26-40e9-a1bc-c52ed17182b1" />

ChrisDoernen committed:
Revert "Programatically set (random) user secrets using dotnet command. Then assert the secrets are available in the application via configuration."

This reverts commit b4dc253ec0ff9eb159b39b465a6850e4f769377f.

ChrisDoernen committed:
Add explicit test class for testing configuration without test appsettings

ChrisDoernen commented:
@Arash-Sabet I reverted my second commit and made a new one with an explicit test case for configuration without appsettings. All tests are green.

ChrisDoernen commented:
@Arash-Sabet what do you think? TestBedFixture line 153 is the main change: We build the configuration always, not only when appsettings are used. This enables using env variables only (without appsettings).

Arash-Sabet commented:
@ChrisDoernen Thanks. I will review it soon.

Arash-Sabet committed:
Update pipeline configs and clean up test code

Bump .NET SDK version to 10.0.101 and increment patch/revision numbers in both Azure pipeline YAML files. Remove unnecessary blank lines in ConfigurationTestsWithoutAppsettings.cs for cleaner test code.

Arash-Sabet committed:
Skip empty filenames in configuration loading

Update GetConfigurationRoot to ignore TestAppSettings entries with null or empty filenames when adding JSON files to the configuration builder.

ChrisDoernen commented:
Hey @Arash-Sabet, thanks so much for your work. I think it is a useful improvement, I will used it right away! 

Happy holidays to you and your family!
